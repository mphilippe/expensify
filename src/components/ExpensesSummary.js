import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import '../config/numeral.js';
import selectExpenses from '../selectors/expenses';
import expensesTotal from '../selectors/expenses-total';

// export const ExpensesSummary = (props) => (
//   <div>
//     {props.expenses.length === 0 ? (
//       <p>No expense</p>
//     ) : (
//       <p>
//         Viewing {props.expenses.length} 
//         expense{props.expenses.length > 1 ? 's' : ''} 
//         totalling 
//         {numeral(expensesTotal(props.expenses) / 100).format('$0,0[.]00')}
//       </p>
//     )}
//   </div>
// );

// const mapStateToProps = (state) => ({
//   expenses: selectExpenses(state.expenses, state.filters)
// });

export const ExpensesSummary = ({ expenseCount, expensesTotal }) => {
  const expenseWord = 'expense' + (expenseCount > 1 ? 's' : '');
  const formattedTotal = numeral(expensesTotal / 100).format('$0,0[.]00');
  return (
    <div>
      {expenseCount === 0 ? (
        <p></p>
      ) : (
        <p>
          Viewing {expenseCount} {expenseWord} totalling {formattedTotal}
        </p>
      )}
    </div>
  );
};

const mapStateToProps = (state) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);
  return {
    expenseCount: visibleExpenses.length,
    expensesTotal: expensesTotal(visibleExpenses)
  };
};

export default connect(mapStateToProps)(ExpensesSummary);
