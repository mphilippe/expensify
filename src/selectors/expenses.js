import moment from 'moment';

export default (expenses, {
  text,
  sortBy,
  startDate,
  endDate
}) => {
  text = text.toLowerCase();
  return expenses.filter((expense) => {
    const createdAt = moment(expense.createdAt);
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAt, 'day') : true;
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAt, 'day') : true;
    const textMatch = text.length === 0 || (expense.note.toLowerCase().includes(text) || expense.description.toLowerCase().includes(text));

    return startDateMatch && endDateMatch && textMatch;
  })
    .sort((a, b) => {
      if (sortBy === 'date') {
        return a.createdAt < b.createdAt ? 1 : -1;
      } else if (sortBy === 'amount') {
        return a.amount < b.amount ? 1 : -1;
      }
    });
};
