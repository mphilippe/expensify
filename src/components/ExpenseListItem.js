import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { removeExpense } from '../actions/expenses';

export const ExpenseListItem = ({ id, description, note, amount, createdAt, dispatch }) => (
  <div>
    <h3><Link to={`/edit/${id}`}>{description}</Link></h3>
    {note && <p>{note}</p>}
    <p>{createdAt} - {amount}</p>
    <button onClick={(e) => {
      dispatch(removeExpense({
        id
      }));
    }}>Remove</button>
  </div>
);

export default connect()(ExpenseListItem);
