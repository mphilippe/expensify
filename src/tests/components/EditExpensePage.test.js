import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses';

let wrapper, editExpense, removeExpense, history;

beforeEach(() => {
  editExpense = jest.fn();
  removeExpense = jest.fn();
  history = {
    push: jest.fn()
  };
  wrapper = shallow(<EditExpensePage
    expense={expenses[0]}
    editExpense={editExpense}
    removeExpense={removeExpense}
    history={history}
  />);
});

test('Should render EditExpensePage properly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('Should handle editExpense correctly', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0]);
  expect(editExpense).toHaveBeenLastCalledWith(expenses[0].id, expenses[0]);
  expect(history.push).toHaveBeenLastCalledWith('/');
});

test('Should handle removeExpense correctly', () => {
  wrapper.find('button').simulate('click');
  expect(removeExpense).toHaveBeenLastCalledWith(expenses[0].id);
  expect(history.push).toHaveBeenLastCalledWith('/');
});
