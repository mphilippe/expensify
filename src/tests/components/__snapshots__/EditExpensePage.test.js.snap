// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render EditExpensePage properly 1`] = `
<div>
  <h1>
    Edit expense
  </h1>
  <ExpenseForm
    expense={
      Object {
        "amount": 6000,
        "createdAt": 0,
        "description": "coffee",
        "id": "aaa",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
